### PR TITLE
Adds support for using primitive types in joins.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Fixes: https://github.com/confluentinc/ksql/issues/4094

BREAKING CHANGE: Some existing joins may now fail and the type of `ROWKEY` in the result schema of joins may have changed.

When `ROWKEY` was always a `STRING` it was possible to join an `INTEGER` column with a `BIGINT` column.  This is no longer the case. A `JOIN` requires the join columns to be of the same type. (See https://github.com/confluentinc/ksql/issues/4130 which tracks adding support for being able to `CAST` join criteria).

Where joining on two `INT` columns would previously have resulted in a schema containing `ROWKEY STRING KEY`, it would not result in `ROWKEY INT KEY`.

*NOTE*: a fair few tests needed to be fixed as they were now doing invalid joins or expecting a STRING key.  I also removed some tests that were failing as they duplicated existing QTT tests.

### Testing done 

usual

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 1]
```diff
@@ -252,37 +243,7 @@ Joiner getJoiner(final DataSourceType leftType,
...
-        );
-      }
-
-      throw new KsqlException(
-          "Source table (" + tableName.toString(FormatOptions.noEscape()) + ") key column ("
-              + keyColumn.get().ref().toString(FormatOptions.noEscape()) + ") "
-              + "is not the column used in the join criteria ("
-              + joinFieldName.toString(FormatOptions.noEscape()) + "). "
-              + "Only the table's key column or 'ROWKEY' is supported in the join criteria."
-      );
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java]: These checks are now done from the constructor.

File: ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java [hunk 2]
```diff
@@ -421,33 +413,6 @@ public void testSelectWithExpression() {
 
   }
 
-  @Test
```
Comment (vpapavas) [inline ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java]: Also, why did you remove these tests?

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 3]
```diff
@@ -465,4 +423,80 @@ private static DataSourceType calculateSinkType(
...
+    // Change of key columns required
+
+    if (source.getDataSourceType() == DataSourceType.KTABLE) {
+      // Tables do not support rekey:
+      final String sourceName = source.getDataSource().getName().toString(FormatOptions.noEscape());
+
+      if (!keyColumn.isPresent()) {
+        throw new KsqlException(
+            "Invalid join criteria: Source table (" + sourceName + ") has no key column "
+                + "defined. Only 'ROWKEY' is supported in the join criteria."
```
Comment (agavra) [inline ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java]:
```suggestion
                + "defined. Only 'ROWKEY' is supported in the join criteria for a TABLE."
```

**Comment by agavra** [comment]
> Currently, it fails if the columns are not of the same type.

@vpapavas - I think this is OK for now, I think it makes sense to require explicit casting for non-equivalent types of fields.

**Comment by big-andy-coates** [comment]
@vpapavas 

> Hey @big-andy-coates ! I think it LGTM.
> A couple of questions:
> 
> 1. Do we want to test with other data types as well besides `BIGINT`?

I'll ensure that's done.

> 2. What is the decision regarding casting? Currently, it fails if the columns are not of the same type.

There's an issue tracking that work: https://github.com/confluentinc/ksql/issues/4130

Thanks for the review!

File: ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java]: Because there are QTT tests that cover the same thing.

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java]: Good call. done.
